src/data.rs

```diff
@@ -0,0 +1,137 @@
+use super::utils::collections::{btree_map, BTreeMap, BTreeSet};
+use core::{cell::RefCell, iter::Chain};
+
+/// A [DataRecorder] that records read requests to the underlying key-value map.
+/// The data recorder is used to generate a proof for read requests. The proof is used to
+/// enable stateless execution of programs on the Miden virtual machine.
+///
+/// The [DataRecorder] is composed of three parts:
+/// - `init`: which contains the initial key-value pairs from the underlying data set.
+/// - `new`: which contains key-value pairs which are created during transaction execution.
+/// - `recorder`: which contains the keys from the initial data set that are read during
+///   program execution.
+#[derive(Debug, Clone, Eq, PartialEq)]
+pub struct DataRecorder<K, V> {
+    init: BTreeMap<K, V>,
+    new: BTreeMap<K, V>,
+    recorder: RefCell<BTreeSet<K>>,
+}
+
+impl<K: Ord + Clone, V: Clone> DataRecorder<K, V> {
+    // CONSTRUCTOR
+    // --------------------------------------------------------------------------------------------
+    /// Returns a new [DataRecorder] instance initialized with the provided key-value pairs
+    /// ([BTreeMap]).
+    pub fn new(init: BTreeMap<K, V>) -> Self {
```

Comment:
Do we want to use a more generic type for the constructor, something like `IntoIterator<Item = (K, V)>`